src/psyclone/tests/psyir/transformations/matmul2code_trans_test.py

```diff
@@ -0,0 +1,389 @@
+            in str(excinfo.value))
+
+
+def test_validate8():
+    '''Check that the Matmul2Code validate method raises the expected
+    exception when the supplied node is a MATMUL binary operation but
+    the third (or higher) dimension of the first (matrix) argument is
+    indexed via a range.
+
+    '''
+    trans = Matmul2CodeTrans()
+    matmul = create_matmul()
+    matrix = matmul.children[0]
+    my_range = matrix.children[0]
+    matrix.children[2] = my_range
+    with pytest.raises(GenerationError) as excinfo:
+        trans.validate(matmul)
+    assert ("Generation Error: To use matmul2code_trans on matmul, indices 2 "
+            "onwards of the first (matrix) argument should not be Ranges but "
+            "found Range at index 2." in str(excinfo.value))
+
+
+def test_validate9():
+    '''Check that the Matmul2Code validate method raises the expected
+    exception when the supplied node is a MATMUL binary operation but
+    the first dimension of its second (vector) argument is not a full
+    range.
+
+    '''
+    trans = Matmul2CodeTrans()
+    matmul = create_matmul()
+    vector = matmul.children[1]
+    vector.children[0] = Literal("1", DataType.INTEGER)
+    with pytest.raises(GenerationError) as excinfo:
+        trans.validate(matmul)
+    assert ("Generation Error: To use matmul2code_trans on matmul, index 0 "
+            "of the 2nd (vector) argument 'x' must be a full range."
+            in str(excinfo.value))
+
+
+def test_validate10():
+    '''Check that the Matmul2Code validate method raises the expected
+    exception when the supplied node is a MATMUL binary operation but
+    the second (or higher) dimension of the second (vector) argument is
+    indexed via a range.
+
+    '''
+    trans = Matmul2CodeTrans()
+    matmul = create_matmul()
+    vector = matmul.children[1]
+    my_range = vector.children[0]
+    vector.children[1] = my_range
+    with pytest.raises(GenerationError) as excinfo:
+        trans.validate(matmul)
+    assert ("Generation Error: To use matmul2code_trans on matmul, indices 1 "
+            "onwards of the second (vector) argument should not be Ranges "
+            "but found Range at index 1." in str(excinfo.value))
+
+
+def test_validate11():
+    '''Check that the Matmul2Code validate method returns without any
+    exceptions when the supplied node is a MATMUL binary operation
+    that obeys the required rules and constraints.
+
+    '''
+    trans = Matmul2CodeTrans()
+    matmul = create_matmul()
+    trans.validate(matmul)
+
+
+def test_apply1():
+    '''Test that the matmul2code apply method produces the expected
+    PSyIR. We use the Fortran backend to help provide the test for
+    correctness. This example includes extra indices for the vector
+    and matrix arrays with these indices being references.
+
+    '''
+    trans = Matmul2CodeTrans()
+    matmul = create_matmul()
+    trans.apply(matmul)
+    writer = FortranWriter()
+    result = writer(matmul.root)
```

Comment:
Please add a check that this code compiles (which it won't currently).

Reply:
It's rather embarrassing that I forgot to test for compilation. However, I've now added some. I find they do compile OK - were you saying that they didn't for you?

Reply:
Not to worry - that's why we have code review :-)